### DISCUSSION TITLE
Retrieve number of dates inside a given month

### DISCUSSION THREAD

**Discussion by mainrs** [time-rs/time#321] [Question] [answered]
Is there a way to get the number of days within a given month? Not sure if this is out-of-scope for the library but I was hoping for an enum named `Month` that can be used to get the number of days, get the enum variant from an integer, maybe even get the next or previous month for a given month.

If this is in-scope and just not part of the API, I'd be happy to create a PR :) 

__Edit:__ I just saw that there is a weekday enum already, maybe this is in-scope after all.

**Comment by jhpratt** [comment]
There's currently no _public_ API to do this. There is an internal method that does exactly this, but it relies on valid input (it panics if invalid). I've considered adding a method that returns `Option<u8>`, but that just feels odd to me.

I have looked into creating an `enum Month` in the past, and determined that it wasn't worth it. Months, unlike weekdays, have a canonical numerical representation, so it's widely understood that `1` means "January".

**Reply by mainrs** [reply]
> I have looked into creating an enum Month in the past, and determined that it wasn't worth it. Months, unlike weekdays, have a canonical numerical representation, so it's widely understood that 1 means "January".

I mean this could be encoded as well:

```rust
enum Month {
    January = 1,
    February = 2,
}
```

I just don't think that this is a feature that should go into upstream crates and would rather fit into the time crate itself :thinking:

**Reply by jhpratt** [reply]
Yes, but having an enum like that would lead to the logical conclusion that it should be used everywhere, not just for a helper method. I could, of course, have a private method that allows creating the enum from the numerical representation, but doing so would require the compiler to add additional branches where none currently exist — it's not possible to (safely, stably and in a `const fn`) let the compiler know that the result of the a calculation is guaranteed to be in a certain range of values.

Once `const_panic` lands on stable (it's not yet stabilized on nightly), it's something to _consider_, as I could use some tricks like stdlib uses to make the panic "unlikely" and never inlined. But aside from that, doing this would almost certainly have notable performance implications.

And just to state the obvious, but even with `enum Month`, you still need to know what the year is to know how many days are in the month; leap years are a thing.

**Reply by mainrs** [reply]
> Once const_panic lands on stable (it's not yet stabilized on nightly), it's something to consider, as I could use some tricks like stdlib uses to make the panic "unlikely" and never inlined. But aside from that, doing this would almost certainly have notable performance implications.

I wasn't aware that you need this inside of `const` context. It's true that without that feature you probably can't implement it.

> And just to state the obvious, but even with `enum Month`, you still need to know what the year is to know how many days are in the month; leap years are a thing.

That's true! Although AFAIK calculating the leap year is just some modulo magic which can be done inside of `const` context.

**Reply by jhpratt** [reply]
If you look at the docs for the main branch, just about everything is `const fn` within compiler limits. I've made significant efforts for this to be the case, so I'd certainly prefer to avoid a regression.

Calculating if it's a leap year is trivial! It's already exposed via a utility method. I just mentioned that because even if you had `enum Month`, that's insufficient information to handle the general case for what you're trying to do.

**Comment by jhpratt** [comment, marked as answer]
Just FYI there is a `Month` enum in the newly-release v0.3.0.
